### ISSUE TITLE
trial fails to run on Windows with Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8577]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)| @rodrigc reported|
|-|-|
|Trac ID|trac#8577|
|Type|defect|
|Created|2016-07-09 09:07:42Z|
|Branch|https://github.com/twisted/twisted/tree/8577-rodrigc-trial-win32-py3|

If Python 3 is used on Windows, then trial fails to start.
This invocation works with Python 2, but fails with Python 3:

```
C:\Users\Craig\twisted>python bin/trial twisted
Traceback (most recent call last):
  File "bin/trial", line 22, in <module>
    run()
  File "C:\Users\Craig\twisted\twisted\scripts\trial.py", line 616, in run
    test_result = trialRunner.run(suite)
  File "C:\Users\Craig\twisted\twisted\trial\runner.py", line 959, in run
    return self._runWithoutDecoration(test, self._forceGarbageCollection)
  File "C:\Users\Craig\twisted\twisted\trial\runner.py", line 983, in _runWithoutDecoration
    oldDir = self._setUpTestdir()
  File "C:\Users\Craig\twisted\twisted\trial\runner.py", line 886, in _setUpTestdir
    testdir, self._testDirLock = util._unusedTestDirectory(base)
  File "C:\Users\Craig\twisted\twisted\trial\util.py", line 365, in _unusedTestDirectory
    testdir.child('_trial_marker').setContent(b'')
  File "C:\Users\Craig\twisted\twisted\python\filepath.py", line 1536, in setContent
    os.rename(sib.path, self.path)
ValueError: rename: src and dst must be the same type
```

<details><summary>Searchable metadata</summary>

```
trac-id__8577 8577
type__defect defect
reporter__rodrigc rodrigc
priority__normal normal
milestone__Python_3_x Python-3.x
branch__8577_rodrigc_trial_win32_py3 8577-rodrigc-trial-win32-py3
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__None None
time__1468055262501376 1468055262501376
changetime__1468715099897402 1468715099897402
version__None None
owner__rodrigc rodrigc

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_2"></a>|@rodrigc commented|
|-|-|

See:
https://github.com/twisted/twisted/pull/303

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @rodrigc|
|-|-|

Thanks for the patch!

I have triggered the unsupported windows builder for this branch

https://buildbot.twistedmatrix.com/boxes-unsupported?branch=8577-rodrigc-trial-win32-py3&num_builds=10


------

The tests are very slow.

Regarding the topfile

```
trial can now run some tests on Windows with Python 3
```

can you be more specific regarding what are those "some tests" ? Which tests can be run?

This note will be part of the next release notes so I think that this could be a valid questions :)


-------

I don't understand why we need to force the bytes path here?

AFAIK, on windows all paths are Unicode and on top of that filepath was updated to work with both unicode and bytes paths.

Isn't this a bug in filepath ?

--------

py35 builder is working and now I am working an updating the py34 builder.


Please check my comment and resubmit.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

Python 3.4 build is also running https://buildbot.twistedmatrix.com/builders/win2012r2-py3.4/builds/1 :)

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_5"></a>|@rodrigc removed owner|
|-|-|

No, I am not going to be more specific in the topfile.
There are a lot of things that need to be fixed to get the tests working on Python 3.
Anyone who is interested can look at the buildbot or run the tests themselves on Python 3.

Regarding FilePath, it does not transparently handle Unicode vs. bytes issues.
The caller of FilePath needs to make sure they use FilePath.asBytesMode() or FilePath.asTextMode(), and also to pass in a string or bytes as appropriate.

A lot more work needs to be done to fix up trial, but this one-liner at least
allows the tests to run on Windows with Python 3.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_6"></a>|@Lukasa set owner to @rodrigc|
|-|-|

Yup, this seems exactly as messy as I'd expect. I'm with rodrigc here, I don't think there's very much that can meaningfully be done to clean up the topfile here. Merge when ready.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|GitHub set status to `closed`|
|-|-|

In changeset ec71728efdd9b37775458d668a918067f39fb441
```
#!CommitTicketReference repository="" revision="ec71728efdd9b37775458d668a918067f39fb441"
Merge pull request #303 from twisted/8577-rodrigc-trial-win32-py3

Author: rodrigc
Reviewer: Lukasa
Fixes: #8577

Fix trial on Windows with Python 3
```

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)|@rodrigc set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|GitHub set status to `closed`|
|-|-|

In changeset dae42dcba15e0ecc5b16f3d104b52cf7ca2f4051
```
#!CommitTicketReference repository="" revision="dae42dcba15e0ecc5b16f3d104b52cf7ca2f4051"
Merge pull request #342 from twisted/8577-rodrigc-win-trial-2

Author: rodrigc
Reviewer: markrwilliams
Fixes: #8577

Remove "python" before invoking tests via "tox -e py35-tests"
```
